trie/triedb/pathdb/asyncnodebuffer.go

```diff
@@ -0,0 +1,422 @@
+	}
+	var (
+		delta         int64
+		overwrite     int64
+		overwriteSize int64
+	)
+	for owner, subset := range nodes {
+		current, exist := nc.nodes[owner]
+		if !exist {
+			// Allocate a new map for the subset instead of claiming it directly
+			// from the passed map to avoid potential concurrent map read/write.
+			// The nodes belong to original diff layer are still accessible even
+			// after merging, thus the ownership of nodes map should still belong
+			// to original layer and any mutation on it should be prevented.
+			current = make(map[string]*trienode.Node)
+			for path, n := range subset {
+				current[path] = n
+				delta += int64(len(n.Blob) + len(path))
+			}
+			nc.nodes[owner] = current
+			continue
+		}
+		for path, n := range subset {
+			if orig, exist := current[path]; !exist {
+				delta += int64(len(n.Blob) + len(path))
+			} else {
+				delta += int64(len(n.Blob) - len(orig.Blob))
+				overwrite++
+				overwriteSize += int64(len(orig.Blob) + len(path))
+			}
+			current[path] = n
+		}
+		nc.nodes[owner] = current
+	}
+	nc.updateSize(delta)
+	nc.layers++
+	gcNodesMeter.Mark(overwrite)
+	gcBytesMeter.Mark(overwriteSize)
+	return nil
+}
+
+func (nc *nodecache) updateSize(delta int64) {
+	size := int64(nc.size) + delta
+	if size >= 0 {
+		nc.size = uint64(size)
+		return
+	}
+	s := nc.size
+	nc.size = 0
+	log.Error("Invalid pathdb buffer size", "prev", common.StorageSize(s), "delta", common.StorageSize(delta))
+}
+
+func (nc *nodecache) reset() {
+	atomic.StoreUint64(&nc.immutable, 0)
+	nc.layers = 0
+	nc.size = 0
+	nc.nodes = make(map[common.Hash]map[string]*trienode.Node)
+}
+
+func (nc *nodecache) empty() bool {
+	return nc.layers == 0
+}
+
+func (nc *nodecache) flush(db ethdb.KeyValueStore, clean *fastcache.Cache, id uint64) error {
+	if atomic.LoadUint64(&nc.immutable) != 1 {
+		return errFlushMutable
+	}
+
+	// Ensure the target state id is aligned with the internal counter.
+	head := rawdb.ReadPersistentStateID(db)
+	if head+nc.layers != id {
+		return fmt.Errorf("buffer layers (%d) cannot be applied on top of persisted state id (%d) to reach requested state id (%d)", nc.layers, head, id)
+	}
+	var (
+		start = time.Now()
+		batch = db.NewBatchWithSize(int(float64(nc.size) * DefaultBatchRedundancyRate))
+	)
+	nodes := writeNodes(batch, nc.nodes, clean)
+	rawdb.WritePersistentStateID(batch, id)
+
+	// Flush all mutations in a single batch
+	size := batch.ValueSize()
+	if err := batch.Write(); err != nil {
+		return err
+	}
+	commitBytesMeter.Mark(int64(size))
+	commitNodesMeter.Mark(int64(nodes))
+	commitTimeTimer.UpdateSince(start)
+	log.Debug("Persisted pathdb nodes", "nodes", len(nc.nodes), "bytes", common.StorageSize(size), "elapsed", common.PrettyDuration(time.Since(start)))
+	nc.reset()
+	return nil
+}
+
+func (nc *nodecache) merge(nc1 *nodecache) (*nodecache, error) {
+	if nc == nil && nc1 == nil {
+		return nil, nil
+	}
+	if nc == nil || nc.empty() {
+		atomic.StoreUint64(&nc1.immutable, 0)
```

Comment:
if background is flushing
current is nil
now merge them, then set background. immutable to false?

Reply:
fixed